incubator/virtualcluster/pkg/syncer/patrol/differ/differ.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differ
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+)
+
+type ClusterObject struct {
+	metav1.Object
+	Cluster string
+}
+
+func (c ClusterObject) String() string {
+	var ns = c.GetNamespace()
+	if c.Cluster != "" {
+		ns = conversion.ToSuperMasterNamespace(c.Cluster, ns)
+	}
+	return fmt.Sprintf("%s%c%s", ns, '/', c.GetName())
+}
+
+func (c ClusterObject) GetCluster() string {
+	return c.Cluster
+}
+
+type Handler interface {
+	OnAdd(obj ClusterObject)
+	OnDelete(obj ClusterObject)
+	OnUpdate(obj1, obj2 ClusterObject)
+}
+
+type Differ interface {
+	Insert(object ...ClusterObject)
+	Has(object ClusterObject) bool
+	Get(key string) ClusterObject
+	Len() int
+	Delete(object ClusterObject)
+	Clear()
+	KeySet() sets.String
+	// Difference compute the different keys between caller and callee.
+	// receive a handler and execute to keep consistent with caller.
+	Difference(Differ, Handler)
+}
+
+type container struct {
+	set map[string]ClusterObject
+}
+
+func NewDiffSet(objects ...ClusterObject) Differ {
+	c := &container{
+		set: make(map[string]ClusterObject),
+	}
+	c.Insert(objects...)
+	return c
+}
+
+func (c *container) Get(key string) ClusterObject {
+	return c.set[key]
+}
+
+func (c *container) Len() int {
+	return len(c.set)
+}
+
+func (c *container) Insert(objects ...ClusterObject) {
+	for _, object := range objects {
+		c.set[object.String()] = object
+	}
+}
+
+func (c *container) Has(object ClusterObject) bool {
+	_, contained := c.set[object.String()]
+	return contained
+}
+
+func (c *container) KeySet() sets.String {
+	s := sets.NewString()
+	for k := range c.set {
+		s.Insert(k)
+	}
+	return s
+}
+
+func (c *container) Delete(object ClusterObject) {
+	delete(c.set, object.String())
+}
+
+func (c *container) Clear() {
+	c.set = make(map[string]ClusterObject)
+}
+
+func (c *container) Difference(set2 Differ, handler Handler) {
+	keySet1 := c.KeySet()
+	keySet2 := set2.KeySet()
+	intersection := sets.NewString()
+
+	for k := range keySet1 {
+		if keySet2.Has(k) {
+			intersection.Insert(k)
+			keySet1.Delete(k)
+			keySet2.Delete(k)
+		}
+	}
+
+	for k := range keySet1 {
```

Comment:
Can we use WaitGroup to call multiple handler.OnAdd in parallel. The same for handler.OnUpdate and Handler.OnDelete?

Reply:
good idea, update